pkg/ctl/install/install.go

```diff
@@ -0,0 +1,435 @@
+		logger.Info("Creating namespace %s", fi.opts.templateParams.Namespace)
+		if err:= createNamespaceSynchronously(fi.k8sClientSet, fi.opts.templateParams.Namespace); err != nil {
+			return err
+		}
+	}
+
+	logger.Info("Installing Flux into the cluster")
+	if err := fi.applyManifestsInDir(ctx, fluxManifestDir); err != nil {
+		return err
+	}
+
+	logger.Info("Waiting for Flux to start")
+	fluxSSHKey, err :=  waitForFluxToStart(ctx, fi.opts, fi.k8sRestConfig, fi.k8sClientSet)
+	if err != nil {
+		return err
+	}
+	logger.Info("Flux started successfully")
+
+	logger.Info("Committing and pushing Flux manifests to %s", fi.opts.templateParams.GitURL)
+	if err := fi.addCommitAndPushFluxManifests(ctx, cloneDir); err != nil {
+		return err
+	}
+	cleanCloneDir = true
+
+	logger.Info("In order for Flux to operate properly make sure it has SSH access to %s",
+		fi.opts.templateParams.GitURL)
+	logger.Info("Flux's current Public SSH key is:\n%s", fluxSSHKey.Key)
+
+	return nil
+}
+
+func (fi *fluxInstaller) cloneRepo(ctx context.Context) (string, error){
+	cloneDir, err := ioutil.TempDir(os.TempDir(), "eksctl-install-flux-clone")
+	if err != nil {
+		return "", fmt.Errorf("cannot create temporary directory: %s", err)
+	}
+	cloneCtx, cloneCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer cloneCtxCancel()
+	args := []string{"clone", "-b", fi.opts.templateParams.GitBranch, fi.opts.templateParams.GitURL, cloneDir}
+	err = runGitCmd(cloneCtx,cloneDir, args...)
+	return cloneDir, err
+}
+
+func (fi *fluxInstaller) addCommitAndPushFluxManifests(ctx context.Context, cloneDir string) error {
+	// Add
+	addCtx, addCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer addCtxCancel()
+	if err := runGitCmd(addCtx, cloneDir, "add", "--", fi.opts.gitFluxPath); err != nil {
+		return err
+	}
+
+	// Confirm there is something to commit, otherwise move on
+	diffCtx, diffCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer diffCtxCancel()
+	if err := runGitCmd(diffCtx, cloneDir, "diff", "--cached", "--quiet", "--", fi.opts.gitFluxPath); err == nil {
+		logger.Info("Nothing to commit (the repository contained identical manifests), moving on")
+		return nil
+	} else if  _, ok := err.(*exec.ExitError); !ok{
+		return err
+	}
+
+	// Commit
+	commitCtx, commitCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer commitCtxCancel()
+	args := []string{"commit",
+		"-m", "Add Initial Flux configuration",
+		fmt.Sprintf("--author=%s <%s>", fi.opts.templateParams.GitUser, fi.opts.templateParams.GitEmail),
+	}
+	if err := runGitCmd(commitCtx, cloneDir, args...); err != nil {
+		return err
+	}
+
+	// Push
+	pushCtx, pushCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer pushCtxCancel()
+	return runGitCmd(pushCtx, cloneDir, "push")
+}
+
+func (fi *fluxInstaller) applyManifestsInDir(ctx context.Context, dir string) error {
+	// NOTE: invoking kubectl is not ideal, but it's the less hairy way to do it until
```

Comment:
We have [`CreateOrUpdate`](https://godoc.org/github.com/weaveworks/eksctl/pkg/kubernetes#RawResource.CreateOrReplace), which should be sufficient for this purpose.

I am okay with `kubectl` being used while it's experimental, but long-term I'd rather avoid introducing dependency on `kubectl`.

Reply:
Having said that, switching to `CreateOrUpdate` should be very easy. And you won't have to write out any temporary files. It's also testable with fake client.

Reply:
After looking around in the codebase to see how to use these utilities, and a few experiments, I came up with this @errordeveloper: https://github.com/weaveworks/eksctl/pull/1064/commits/99a939cb755cb55642c286dd8df19ddbb9ee7b2a
Based on my manual end-to-end testing, this seems to work. However, is it what you had in mind?

Reply:
I am all in for removing the kubectl dependency, but let's make sure that `kubectl apply` works on changes to the resource manifests later on (otherwise Flux won't work on the resources).

Reply:
Latest revision of this PR removes `kubectl` and `kubectl apply` from inside Flux seems to still work afterwards, i.e. https://github.com/marccarre/my-gitops-repo/commit/88b7eb0dd4d315f4a657e937b0fb06447b3310be got applied:
**Installation of Flux**:
```console
$ EKSCTL_EXPERIMENTAL=true ./eksctl -f examples/01-simple-cluster.yaml install flux --git-url=git@github.com:marccarre/my-gitops-repo.git --git-email=carre.marc@gmail.com
[ℹ]  Cloning git@github.com:marccarre/my-gitops-repo.git
Cloning into '/var/folders/24/d3mml6bn20nftpt91cfldq1h0000gn/T/eksctl-install-flux-clone968027774'...
remote: Enumerating objects: 15, done.
remote: Counting objects: 100% (15/15), done.
remote: Compressing objects: 100% (13/13), done.
remote: Total 15 (delta 0), reused 13 (delta 0), pack-reused 0
Receiving objects: 100% (15/15), 6.15 KiB | 6.15 MiB/s, done.
[ℹ]  Writing Flux manifests
[ℹ]  Creating namespace flux
[ℹ]  Installing Flux into the cluster
[ℹ]  created "flux:ServiceAccount/flux"
[ℹ]  created "ClusterRole.rbac.authorization.k8s.io/flux"
[ℹ]  created "ClusterRoleBinding.rbac.authorization.k8s.io/flux"
[ℹ]  created "flux:Deployment.apps/flux"
[ℹ]  created "flux:Secret/flux-git-deploy"
[ℹ]  created "flux:Deployment.apps/memcached"
[ℹ]  created "flux:Service/memcached"
[ℹ]  replaced "Namespace/flux"
[ℹ]  Waiting for Flux to start
[ℹ]  Flux started successfully
[ℹ]  Committing and pushing Flux manifests to git@github.com:marccarre/my-gitops-repo.git
[master ad1e657] Add Initial Flux configuration
 Author: Flux <carre.marc@gmail.com>
 6 files changed, 255 insertions(+)
 create mode 100644 flux/flux-account.yaml
 create mode 100644 flux/flux-deployment.yaml
 create mode 100644 flux/flux-secret.yaml
 create mode 100644 flux/memcache-dep.yaml
 create mode 100644 flux/memcache-svc.yaml
 create mode 100644 flux/namespace.yaml
Enumerating objects: 10, done.
Counting objects: 100% (10/10), done.
Delta compression using up to 8 threads
Compressing objects: 100% (9/9), done.
Writing objects: 100% (9/9), 3.89 KiB | 3.89 MiB/s, done.
Total 9 (delta 0), reused 8 (delta 0)
To github.com:marccarre/my-gitops-repo.git
   882c45f..ad1e657  master -> master
[ℹ]  Flux will operate properly only once it has SSH access to: git@github.com:marccarre/my-gitops-repo.git
[ℹ]  please add the following Flux public SSH key to your repository:
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDJmmIg7dE6RAHfwtlBCtKBUm+eNCNa/lqygcNzSFBjtPrMwqLbuPOQNuiDe+iqIkx0VCokTlF/mk8Fd6zVpVMIiuzJmBvBIO0JQC67V97/P3AAYgzeewuK6iVZ8C9TAtPyQN2/XEsmqizBjj2RdinbxCJjCoDHMSb+pmyVEgzGYasuRtEm89/CkP8mfu9wP0BQ4IymVrAAj6n5MUODqPg5hhnCoe0srSm2B6PDv8VEqeHPN+eWhWGk8cksIYAsis/S4jIDrhCdh1hl02SJV2FoOnAMk3uRIAwTVzoQvz4HRbYgl9x6WrovGJdr1QcKWby7PEMUU4vgwNG6Re558U7t
```
**Before manifest update**:
```console
$ kubectl -n flux get po
NAME                         READY   STATUS    RESTARTS   AGE
flux-6465c6cdb8-szzp9        1/1     Running   0          56s
memcached-75f65589d7-r6dws   1/1     Running   0          56s
```
```console
$ kubectl -n flux describe po memcached-75f65589d7-r6dws
[...]
    Args:
      -m 512
[...]
```
**After manifest update** 
```console
$ kubectl -n flux logs flux-6465c6cdb8-szzp9
[...]
ts=2019-07-31T02:33:02.221160927Z caller=loop.go:111 component=sync-loop event=refreshed url=git@github.com:marccarre/my-gitops-repo.git branch=master HEAD=88b7eb0dd4d315f4a657e937b0fb06447b3310be
ts=2019-07-31T02:33:04.183423651Z caller=sync.go:479 method=Sync cmd=apply args= count=8
ts=2019-07-31T02:33:04.598690069Z caller=sync.go:545 method=Sync cmd="kubectl apply -f -" took=415.205166ms err=null output="namespace/flux configured\nserviceaccount/flux configured\nclusterrole.rbac.authorization.k8s.io/flux configured\nservice/memcached configured\nclusterrolebinding.rbac.authorization.k8s.io/flux configured\nsecret/flux-git-deploy configured\ndeployment.apps/flux configured\ndeployment.apps/memcached configured"
ts=2019-07-31T02:33:08.063391885Z caller=daemon.go:651 component=daemon event="Sync: 11d6d76..88b7eb0, <cluster>:clusterrole/flux, <cluster>:clusterrolebinding/flux, <cluster>:namespace/flux, flux:deployment/flux, flux:deployment/memcached, flux:secret/flux-git-deploy, flux:service/memcached, flux:serviceaccount/flux" logupstream=false
ts=2019-07-31T02:33:11.911992672Z caller=loop.go:206 component=sync-loop tag=flux old= new=88b7eb0dd4d315f4a657e937b0fb06447b3310be
ts=2019-07-31T02:33:13.621547417Z caller=loop.go:111 component=sync-loop event=refreshed url=git@github.com:marccarre/my-gitops-repo.git branch=master HEAD=88b7eb0dd4d315f4a657e937b0fb06447b3310be
```
```console
$ kubectl -n flux get po
NAME                         READY   STATUS    RESTARTS   AGE
flux-6465c6cdb8-szzp9        1/1     Running   0          7m53s
memcached-5585d5d5fb-qn858   1/1     Running   0          21s
```
```console
$ kubectl -n flux describe po memcached-5585d5d5fb-qn858
[...]
    Args:
      -m 1024
[...]
```